### PR TITLE
*: split cluster into a package

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
The Cluster related logic should be put into another package and there is no need to rely on server.

### What is changed and how it works?
This PR mainly does the following things:
1. move cluster logic to an individual package and use an interface named `Server` to solve dependency problems
2. move part of tests in `cluster_test.go` and `cluster_worker_test.go` into `tests/server/cluster`
3. enhance `mockhbstream`
4. rename `cluster` to `tc` or `rc` when it is necessary
5. add some get/set functions

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by rleungx** [comment]
@nolouch @lhy1024 PTAL

File: tests/server/cluster/cluster_test.go [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+		}
+		resp, err := grpcPDClient.StoreHeartbeat(context.Background(), req)
+		c.Assert(err, IsNil)
+		c.Assert(resp.GetHeader().GetError().GetType(), Equals, pdpb.ErrorType_STORE_TOMBSTONE)
+	}
+}
+
+// Make sure PD will not panic if it start and stop again and again.
+func (s *clusterTestSuite) TestRaftClusterRestart(c *C) {
+	tc, err := tests.NewTestCluster(1)
```
Comment (lhy1024) [inline tests/server/cluster/cluster_test.go]: Does it no longer need the previous `cleanup`?

Comment (rleungx) [reply tests/server/cluster/cluster_test.go]: I think we can use `Destroy` to clean up.

File: server/schedule/operator_controller.go [hunk 2]
```diff
@@ -52,18 +52,13 @@ var (
 	StoreBalanceBaseTime float64 = 60
 )
 
-// HeartbeatStreams is an interface of async region heartbeat.
-type HeartbeatStreams interface {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedule/operator_controller.go]: There is a dependency problem.

File: server/grpc_service.go [hunk 3]
```diff
@@ -205,24 +206,25 @@ func (s *Server) PutStore(ctx context.Context, request *pdpb.PutStoreRequest) (*
...
 
-	if err := cluster.putStore(store); err != nil {
+	if err := rc.PutStore(store); err != nil {
 		return nil, status.Errorf(codes.Unknown, err.Error())
 	}
 
 	log.Info("put store ok", zap.Stringer("store", store))
-	cluster.OnStoreVersionChange()
+	rc.OnStoreVersionChange()
+	CheckPDVersion(s.scheduleOpt)
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go]: just move it out from `OnStoreVersionChange`

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1501,3 +1521,96 @@ func (checker *prepareChecker) collect(region *core.RegionInfo) {
...
+// GetSchedulers gets all schedulers.
+func (c *RaftCluster) GetSchedulers() map[string]*scheduleController {
+	c.RLock()
+	defer c.RUnlock()
+	return c.coordinator.getSchedulers()
+}
+
+// AddScheduler adds a scheduler.
+func (c *RaftCluster) AddScheduler(scheduler schedule.Scheduler, args ...string) error {
+	c.RLock()
```
Comment (nolouch) [inline server/cluster/cluster.go]: Lock?

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -1501,3 +1521,96 @@ func (checker *prepareChecker) collect(region *core.RegionInfo) {
...
+	defer c.RUnlock()
+	return c.coordinator.addScheduler(scheduler, args...)
+}
+
+// RemoveScheduler removes a scheduler.
+func (c *RaftCluster) RemoveScheduler(name string) error {
+	c.RLock()
```
Comment (nolouch) [inline server/cluster/cluster.go]: Lock.

File: tests/server/cluster/cluster_test.go [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+	c.Assert(err, IsNil)
+
+	err = tc.RunInitialServers(s.ctx)
+	c.Assert(err, IsNil)
+
+	tc.WaitLeader()
+	leaderServer := tc.GetServer(tc.GetLeader())
+	svr := leaderServer.GetServer()
+	rc := cluster.NewRaftCluster(s.ctx, svr.GetClusterRootPath(), svr.ClusterID(), syncer.NewRegionSyncer(svr), svr.GetClient())
+	// Cluster is not bootstrapped.
```
Comment (lhy1024) [inline tests/server/cluster/cluster_test.go]:
```suggestion

```

File: tests/server/cluster/cluster_test.go [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+
```
Comment (lhy1024) [inline tests/server/cluster/cluster_test.go]:
```suggestion
// Cluster is not bootstrapped.
```

**Pull request merged**